File: exercises/python/12_code_generator/code_generator.py
"""
================================================================================
Exercise 12: Simple Code Generator (Target Code from Three-Address Code)
================================================================================

CONCEPT:
    A code generator is the final phase of a compiler that translates
    intermediate representation (three-address code) into target machine
    instructions.  This implementation produces assembly-like instructions
    (MOV, ADD, SUB, MUL, DIV) using a small set of registers (R0-R7).

ALGORITHM:
    For each three-address code statement of the form  result = left op right:
      1. Load `left` into a register (MOV left, Ri).  If `left` is already in
         a register, skip the MOV (register tracking optimisation).
      2. Apply the operation with `right` (OP right, Ri).  If `right` is in a
         register, use that register name instead.
      3. Record that Ri now holds `result`.  If `result` is not a temporary
         (i.e. it is a user variable), also emit MOV Ri, result to store it.

    A register descriptor tracks which variable each register currently holds.
    An address descriptor tracks which register (if any) holds a given variable.
    This allows the generator to avoid redundant loads when an operand is
    already available in a register.

TIME COMPLEXITY:  O(n) where n is the number of TAC instructions.
SPACE COMPLEXITY: O(r + v) where r = registers, v = live variables.

EXAMPLE:
    TAC Input:         Generated Assembly:
    t1 = a + b         MOV a, R0
                        ADD b, R0
    t2 = t1 * c        MUL c, R0
    t3 = d - e         MOV d, R1
                        SUB e, R1
    x  = t2 + t3       ADD R1, R0
                        MOV R0, x
================================================================================
"""

# Map TAC operators to assembly mnemonics
OP_MAP = {
    "+": "ADD",
    "-": "SUB",
    "*": "MUL",
    "/": "DIV",
}

NUM_REGISTERS = 8  # R0 through R7


class RegisterAllocator:
    """
    Manages a pool of registers and tracks which variable each register holds.
    Uses a simple next-available strategy with spill when all registers are in use.
    """

    def __init__(self, num_registers=NUM_REGISTERS):
        self.num_registers = num_registers
        # register_contents[i] = name of variable currently in Ri, or None
        self.register_contents = [None] * num_registers
        # variable_location[name] = register index, or None
        self.variable_location = {}

    def get_register_name(self, index):
        """Return the printable name for register index."""
        return f"R{index}"

    def find_variable(self, name):
        """Return the register index holding `name`, or -1 if not loaded."""
        return self.variable_location.get(name, -1)

    def allocate(self):
        """Return the index of a free register, or evict the first one."""
        # First, look for an empty register
        for i in range(self.num_registers):
            if self.register_contents[i] is None:
                return i
        # All occupied -- evict R0 (simple strategy)
        self._evict(0)
        return 0

    def _evict(self, index):
        """Remove the variable from a register (no spill code emitted here)."""
        old_var = self.register_contents[index]
        if old_var and old_var in self.variable_location:
            del self.variable_location[old_var]
        self.register_contents[index] = None

    def assign(self, index, variable_name):
        """Record that register `index` now holds `variable_name`."""
        # Remove old mapping for this variable if it exists elsewhere
        old_idx = self.find_variable(variable_name)
        if old_idx >= 0:
            self.register_contents[old_idx] = None

        # Remove old variable from this register
        old_var = self.register_contents[index]
        if old_var and old_var in self.variable_location:
            del self.variable_location[old_var]

        self.register_contents[index] = variable_name
        self.variable_location[variable_name] = index

    def reset(self):
        """Clear all register state."""
        self.register_contents = [None] * self.num_registers
        self.variable_location = {}


class CodeGenerator:
    """
    Translates three-address code into register-based assembly instructions.
    """

    def __init__(self):
        self.allocator = RegisterAllocator()
        self.assembly = []

    def emit(self, instruction):
        """Append an assembly instruction string."""
        self.assembly.append(instruction)

    def operand_name(self, name):
        """
        If `name` is currently in a register, return the register name;
        otherwise return the variable/literal name.
        """
        idx = self.allocator.find_variable(name)
        if idx >= 0:
            return self.allocator.get_register_name(idx)
        return name

    def is_temporary(self, name):
        """Temporaries start with 't' followed by digits."""
        return len(name) >= 2 and name[0] == "t" and name[1:].isdigit()

    def generate(self, tac_statements):
        """
        Generate assembly from a list of TAC strings.
        Each string is of the form:  result = left op right
        or a simple copy:            result = source
        """
        self.assembly = []
        self.allocator.reset()

        for statement in tac_statements:
            parts = statement.split()
            result = parts[0]
            # parts[1] is '='
            left = parts[2]

            if len(parts) == 3:
                # Simple copy: result = source
                src_reg = self.allocator.find_variable(left)
                if src_reg >= 0:
                    # Source already in a register
                    reg_idx = src_reg
                else:
                    reg_idx = self.allocator.allocate()
                    self.emit(f"MOV {left}, {self.allocator.get_register_name(reg_idx)}")
                self.allocator.assign(reg_idx, result)

                if not self.is_temporary(result):
                    self.emit(f"MOV {self.allocator.get_register_name(reg_idx)}, {result}")

            elif len(parts) == 5:
                # Binary operation: result = left op right
                operator = parts[3]
                right = parts[4]
                mnemonic = OP_MAP.get(operator, "???")

                # Load left operand
                left_reg = self.allocator.find_variable(left)
                if left_reg >= 0:
                    reg_idx = left_reg
                else:
                    reg_idx = self.allocator.allocate()
                    self.emit(f"MOV {left}, {self.allocator.get_register_name(reg_idx)}")

                # Apply operation with right operand
                right_name = self.operand_name(right)
                self.emit(f"{mnemonic} {right_name}, {self.allocator.get_register_name(reg_idx)}")

                # Mark register as holding the result
                self.allocator.assign(reg_idx, result)

                # If result is a user variable, store it to memory
                if not self.is_temporary(result):
                    self.emit(f"MOV {self.allocator.get_register_name(reg_idx)}, {result}")

        return self.assembly


def print_assembly(asm_lines):
    """Print assembly instructions with line numbers."""
    for i, line in enumerate(asm_lines, start=1):
        print(f"    {i:3d}:  {line}")


def main():
    print("=" * 65)
    print("   SIMPLE CODE GENERATOR  (Target Code from TAC)")
    print("=" * 65)
    print()

    # ---- Test Case 1 ----
    tac_input_1 = [
        "t1 = a + b",
        "t2 = t1 * c",
        "t3 = d - e",
        "x = t2 + t3",
    ]

    # ---- Test Case 2 ----
    tac_input_2 = [
        "t1 = a + b",
        "t2 = c - d",
        "t3 = t1 * t2",
        "y = t3",
    ]

    # ---- Test Case 3 ----
    tac_input_3 = [
        "t1 = a * b",
        "t2 = c * d",
        "t3 = t1 + t2",
        "t4 = e * f",
        "z = t3 + t4",
    ]

    test_cases = [
        ("x = a + b * c - d / e  (simplified TAC)", tac_input_1),
        ("y = (a + b) * (c - d)", tac_input_2),
        ("z = a*b + c*d + e*f", tac_input_3),
    ]

    generator = CodeGenerator()

    for idx, (description, tac_input) in enumerate(test_cases, start=1):
        print(f"  --- Test Case {idx}: {description} ---")
        print("  Input Three-Address Code:")
        for stmt in tac_input:
            print(f"    {stmt}")
        print()

        asm = generator.generate(tac_input)

        print("  Generated Assembly:")
        print_assembly(asm)
        print()

    print("=" * 65)
    print("  All test cases processed successfully.")
    print("=" * 65)


if __name__ == "__main__":
    main()

File: exercises/cpp/12_code_generator/code_generator.cpp
/*
 * ===========================================================================
 * Exercise 12: Simple Code Generator (Target Code from Three-Address Code)
 *              -- C++ Implementation
 * ===========================================================================
 *
 * CONCEPT:
 *   A code generator translates intermediate representation (three-address
 *   code) into target machine instructions.  This implementation produces
 *   assembly-like instructions (MOV, ADD, SUB, MUL, DIV) using a small pool
 *   of registers (R0-R7).
 *
 * ALGORITHM:
 *   For each TAC statement  result = left op right:
 *     1. If `left` is already in a register, reuse it; otherwise load it
 *        with MOV into a free register.
 *     2. Emit the operation instruction (e.g. ADD right, Ri).
 *     3. Update the register descriptor so Ri holds `result`.
 *     4. If `result` is a user variable (not a temporary), store it back
 *        with MOV Ri, result.
 *   Register and address descriptors eliminate redundant MOV instructions.
 *
 * TIME COMPLEXITY:  O(n) where n = number of TAC instructions.
 * SPACE COMPLEXITY: O(r + v) for registers and variable descriptors.
 *
 * EXAMPLE:
 *   TAC:  t1 = a + b       ->   MOV a, R0  /  ADD b, R0
 *         t2 = t1 * c      ->   MUL c, R0
 *         t3 = d - e       ->   MOV d, R1  /  SUB e, R1
 *         x  = t2 + t3     ->   ADD R1, R0 /  MOV R0, x
 * ===========================================================================
 */

#include <iostream>
#include <string>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <cctype>

static const int NUM_REGISTERS = 8;

/* ---------- helper: operator -> mnemonic --------------------------------- */

static std::string op_to_mnemonic(const std::string &op) {
    if (op == "+") return "ADD";
    if (op == "-") return "SUB";
    if (op == "*") return "MUL";
    if (op == "/") return "DIV";
    return "???";
}

/* ---------- check if a name is a compiler temporary ---------------------- */

static bool is_temporary(const std::string &name) {
    if (name.size() < 2 || name[0] != 't') return false;
    for (size_t i = 1; i < name.size(); ++i)
        if (!std::isdigit(static_cast<unsigned char>(name[i]))) return false;
    return true;
}

/* ---------- TAC statement ------------------------------------------------ */

struct TACStmt {
    std::string result;
    std::string left;
    std::string op;     /* "+", "-", "*", "/", or "" for simple copy */
    std::string right;
};

static TACStmt parse_tac(const std::string &line) {
    TACStmt stmt;
    std::istringstream iss(line);
    std::string eq;
    iss >> stmt.result >> eq >> stmt.left;
    if (iss >> stmt.op >> stmt.right) {
        /* binary operation */
    } else {
        stmt.op.clear();
        stmt.right.clear();
    }
    return stmt;
}

/* ---------- Code Generator ----------------------------------------------- */

class CodeGenerator {
public:
    CodeGenerator() : reg_contents_(NUM_REGISTERS) {}

    std::vector<std::string> generate(const std::vector<std::string> &tac_lines) {
        reset();
        for (const auto &line : tac_lines) {
            TACStmt stmt = parse_tac(line);
            generate_one(stmt);
        }
        return assembly_;
    }

private:
    /* Register descriptor: reg_contents_[i] = variable name or "" */
    std::vector<std::string> reg_contents_;
    /* Address descriptor: variable -> register index (-1 = not loaded) */
    std::unordered_map<std::string, int> var_reg_;
    /* Output buffer */
    std::vector<std::string> assembly_;

    void reset() {
        for (auto &rc : reg_contents_) rc.clear();
        var_reg_.clear();
        assembly_.clear();
    }

    void emit(const std::string &instr) {
        assembly_.push_back(instr);
    }

    int find_var_reg(const std::string &name) const {
        auto it = var_reg_.find(name);
        return (it != var_reg_.end()) ? it->second : -1;
    }

    std::string reg_name(int idx) const {
        return "R" + std::to_string(idx);
    }

    /* Get the best operand name: register if loaded, else the variable name */
    std::string operand(const std::string &name) const {
        int ri = find_var_reg(name);
        return (ri >= 0) ? reg_name(ri) : name;
    }

    int allocate() {
        for (int i = 0; i < NUM_REGISTERS; ++i)
            if (reg_contents_[i].empty()) return i;
        /* Evict R0 */
        evict(0);
        return 0;
    }

    void evict(int ri) {
        if (!reg_contents_[ri].empty()) {
            auto it = var_reg_.find(reg_contents_[ri]);
            if (it != var_reg_.end() && it->second == ri)
                var_reg_.erase(it);
            reg_contents_[ri].clear();
        }
    }

    void assign(int ri, const std::string &name) {
        /* Clear old register for this variable */
        int old_ri = find_var_reg(name);
        if (old_ri >= 0 && old_ri != ri)
            reg_contents_[old_ri].clear();
        /* Clear old variable for this register */
        evict(ri);
        reg_contents_[ri] = name;
        var_reg_[name] = ri;
    }

    void generate_one(const TACStmt &stmt) {
        if (stmt.op.empty()) {
            /* Simple copy: result = source */
            int src_ri = find_var_reg(stmt.left);
            int ri;
            if (src_ri >= 0) {
                ri = src_ri;
            } else {
                ri = allocate();
                emit("MOV " + stmt.left + ", " + reg_name(ri));
            }
            assign(ri, stmt.result);
            if (!is_temporary(stmt.result))
                emit("MOV " + reg_name(ri) + ", " + stmt.result);
        } else {
            /* Binary: result = left op right */
            int left_ri = find_var_reg(stmt.left);
            int ri;
            if (left_ri >= 0) {
                ri = left_ri;
            } else {
                ri = allocate();
                emit("MOV " + stmt.left + ", " + reg_name(ri));
            }
            emit(op_to_mnemonic(stmt.op) + " " + operand(stmt.right) +
                 ", " + reg_name(ri));
            assign(ri, stmt.result);
            if (!is_temporary(stmt.result))
                emit("MOV " + reg_name(ri) + ", " + stmt.result);
        }
    }
};

/* ---------- print helper ------------------------------------------------- */

static void print_assembly(const std::vector<std::string> &asm_lines) {
    for (size_t i = 0; i < asm_lines.size(); ++i)
        std::cout << "    " << (i + 1) << ":  " << asm_lines[i] << "\n";
}

/* ---------- main --------------------------------------------------------- */

int main() {
    std::cout << "=================================================================\n";
    std::cout << "   SIMPLE CODE GENERATOR  (Target Code from TAC)  --  C++\n";
    std::cout << "=================================================================\n\n";

    struct TestCase {
        std::string              description;
        std::vector<std::string> tac;
    };

    std::vector<TestCase> tests = {
        {"x = a + b * c - d / e",
         {"t1 = a + b", "t2 = t1 * c", "t3 = d - e", "x = t2 + t3"}},
        {"y = (a + b) * (c - d)",
         {"t1 = a + b", "t2 = c - d", "t3 = t1 * t2", "y = t3"}},
        {"z = a*b + c*d + e*f",
         {"t1 = a * b", "t2 = c * d", "t3 = t1 + t2",
          "t4 = e * f", "z = t3 + t4"}},
    };

    CodeGenerator generator;

    for (size_t i = 0; i < tests.size(); ++i) {
        std::cout << "  --- Test Case " << (i + 1) << ": "
                  << tests[i].description << " ---\n";
        std::cout << "  Input Three-Address Code:\n";
        for (const auto &line : tests[i].tac)
            std::cout << "    " << line << "\n";
        std::cout << "\n";

        auto asm_out = generator.generate(tests[i].tac);
        std::cout << "  Generated Assembly:\n";
        print_assembly(asm_out);
        std::cout << "\n";
    }

    std::cout << "=================================================================\n";
    std::cout << "  All test cases processed successfully.\n";
    std::cout << "=================================================================\n";
    return 0;
}
